### ISSUE TITLE
DemoActor calling RegisterRemind start to error on each remind

### ISSUE THREAD

**Issue by amanbha** [dapr/dotnet-sdk#177]
Cloning from dapr repo to dotnet-sdk repo
https://github.com/dapr/dapr/issues/864

What version of Dapr?
0.2.x

Expected Behavior
After Registering a reminder, the ReceiveReminder executes without error

Actual Behavior
Solution works great, however, as soon as a reminder is added, the ReceiverReminder fails each time it's invoked. - see attached screen shots of console log

Steps to Reproduce the Problem
Add the line "this.RegisterReminder()" to SaveData, and invoke SaveData once.
RemindFailure
CodeChangeDemoActor

![image](https://user-images.githubusercontent.com/11221966/70095367-6b451f00-15d9-11ea-8590-a5679fae630d.png)
![image](https://user-images.githubusercontent.com/11221966/70095379-71d39680-15d9-11ea-8f72-9b7fe3ce887f.png)

**Comment by amanbha** [comment]
cc @codeputer Thanks for reporting the issue,
I looked at it, the issue is happening because ReminderInfo is serialized synchronously and aspnetcore 3.0 disallows it by default.
I will update the code to async deserializtion.

To unblock, you can change the code where AspNetCore WebHost is created to allow Synchronous IO

Add to `.ConfigureKestrel(options => options.AllowSynchronousIO = true)` to `CreateWebHostBuilder` method so that method will look like:
 

` public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
            WebHost.CreateDefaultBuilder(args)
                .UseStartup<Startup>()
            .ConfigureKestrel(options => options.AllowSynchronousIO = true)
            .UseActors(actorRuntime =>
            {
                actorRuntime.RegisterActor<DemoActor>();
            });`

**Comment by codeputer** [comment]
I tried this in the ConfigureServices....


   public void ConfigureServices(IServiceCollection services)
    {
      services.AddRouting();
      services.Configure<IISServerOptions>(options =>
      {
        options.AllowSynchronousIO = true;
      });
    }

but as far as I can tell, it didn't have any effect

**Comment by amanbha** [comment]
@codeputer Could you please try as I suggested above?

**Comment by codeputer** [comment]
Working on it....

**Comment by codeputer** [comment]
I'm not sure what is going on... but I put in the changes you suggested, and got a 502 back
With the as the payload (I'm using fiddler to post )
HTTP/1.1 502 Fiddler - Connection Failed
Date: Tue, 03 Dec 2019 23:52:14 GMT
Content-Type: text/html; charset=UTF-8
Connection: close
Cache-Control: no-cache, must-revalidate
Timestamp: 16:52:14.245

[Fiddler] The connection to 'localhost' failed. <br />Error: ConnectionRefused (0x274d). <br />System.Net.Sockets.SocketException No connection could be made because the target machine actively refused it 127.0.0.1:3500

I then backed out the changes, to get back to the original error, and I got the same result as above....which means I have to get my environment working like it was prior to your fix... I'll keep you updated unless you can see something here that doesn't make sense to you.

I start the dapr services with this line, attempting to connect to 3500 - which now is NOT working with or without your changes:
[cid:image001.png@01D5A9FB.14661EE0]

So with my resetting of the DAPR environment, and docker resets - something has gone wrong...

Attempting to reset to the previous behaviour with the error I first described to you.. in order to properly see if the changes you suggest will work in that environment.... Just for documentation complete..  I'm using this from fiddler..
[cid:image003.jpg@01D5A9FB.146D98F0]

Thanks for your time... .R

P.s. & Segway: Why would [FromForm] not work on the method signatures when I POST to an Actor?



From: Richard Reukema
Sent: December 3, 2019 4:17 PM
To: dapr/dotnet-sdk <reply@reply.github.com>;

Subject: RE: [dapr/dotnet-sdk] DemoActor calling RegisterRemind start to error on each remind (#177)

Working on it....

**Comment by amanbha** [comment]
@codeputer The screenshots are not attached above, Could you paste them here

**Comment by amanbha** [comment]
@codeputer I have created the PR to fix this https://github.com/dapr/dotnet-sdk/pull/178
You can checkout branch issue_177 and try

**Comment by codeputer** [comment]
Thanks - I had to reset docker and few things, still not quite working right, but I will try the branch! thanks.

**Comment by codeputer** [comment]
My status is at this point ActorClient can invoke DemoActor - which is when I refactored to code to include IRemindable, and things broke.  The fix worked, but my new Actor that I've been testing against is having an issue.  It still unclear to me how the proxy "knows" where the hosting actor in terms of path and port.  Mean time, I'm geting...
HTTP/1.1 500 Internal Server Error
Server: fasthttp
Date: Wed, 04 Dec 2019 06:38:21 GMT
Content-Type: application/json
Content-Length: 111

{"errorCode":"ERR_INVOKE_ACTOR","message":"error finding address for actor type VoiceRequestActor with id abc"}

**Comment by amanbha** [comment]
@codeputer The proxy sends the call to Dapr runtime, Dapr runtime has a map of ActorIds and the Actor Service instances which can serve the requests. The map is populated by a placement service. For more info about it, please see https://github.com/dapr/docs/blob/master/concepts/actor/actor_overview.md#distribution-and-failover

Most probably the placement service is not able to talk to Dapr runtime. Could you please do following to refresh placement service and dapr runtime.:
dapr uninstall
dapr init
Also is it possible for you to share your code?

**Comment by codeputer** [comment]
thanks - I was doing a spike on using Actors with Twilio and what I think is a new way to handle webhooks, and in particular, twilio web hooks.  I'm cleaning up the project as I started it within the dapr branch and got carries away :)   I have to commit to some other work today, but I'll share the repo with  you later... can I send that invite to you directly? I'll need an email via a private channel perhaps?

**Comment by amanbha** [comment]
sure, you can send it to me at amanbha@microsoft.com

**Comment by codeputer** [comment]
I had a good look a that documentation, and I believe I understand how the Actor proxy can find the Actor Implementation (and hence the ActorID) WITHIN a cluster - but what if the Actor client is outside of the cluster?  in my case, I ran the web service as a local IIS service, and it was communicating to the actor in the docker container  Is there something in the web service that points to the placement service of cluster?  (I'm new to docker, so I might be missing something here... I don't understand the development/debug/deployment cycle in visual studio all that well - like how do I step through code in an actor which is running in via dapr in docker?)

**Comment by amanbha** [comment]
@codeputer Actors will be called from other services inside the cluster, so you need another service which has a docker side car running with it and then just use ActorProxy from within that service, it will make calls to Actors through its Dapr side car.

**Comment by codeputer** [comment]
In my code review last night, I found a couple of errors in my code: 1) in my rework of my project structure, I somehow lost the registration of my own actor, and hence placement services won't know anything about it...a more explicit error message in this regard would be interesting.  The magic string in the proxy call is also something that will cause an issue IMO, and is not emphasized enough in the documentation.  I'm going to work on a video of my debugging environment once I get comfortable with my development.  I also found Constant.cs which gave me a clue how my actor communicates with the port that is open with DAPR ( DaprHttpPortEnvironmentVariable ).  I'm also confused with your statement, as my actor instantiation occurs outside of DAPR as I'm running it within Visual Studio from a Web service that invokes actors within DAPR.. more learnings to come I'm sure.... thanks for your help on the journey... how do you run your debugging environment would be helpful.  I have studio open, and several dos windows that all need to work to establish the workflow that I'm doing with Twilio.

**Comment by amanbha** [comment]
@codeputer On your dev machine, the actor proxy call will go to localhost:3500 (by default). So if you had Dapr running on port 3500, it will resolve the ctor correctly.
For missing documentation, please feel free to raise PR or suggest the changes.

**Comment by amanbha** [comment]
@codeputer Can I close this issue, it should be fixed with latest release. We can open new issue to track if you run into problems with it

**Comment by codeputer** [comment]
that fine, please close - thanks very much for your help - I've moved to the nuget packages - which I'm assuming has this fix already.  If not, please let me know.  

Regards,
Richard

**Comment by amanbha** [comment]
Sounds good.
Yes, the latest nuget package has the fix. We also released 0.3.0 version of Dapr today, you can update to it as well https://github.com/dapr/dapr/releases/tag/v0.3.0
